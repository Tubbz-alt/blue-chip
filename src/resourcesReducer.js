import produce from "immer";

const initialState = {};

export default function resourcesReducer(state = initialState, action) {
  const {
    type,
    id,
    attributes,
    links,
    relationships,
    resourcesById,
    resourceTypes,
    resourceType
  } = action;
  return produce(state, draft => {
    switch (type) {
      case "ADD_OR_REPLACE_RESOURCE_BY_ID":
        _initializeResource(draft, resourceType);

        draft[resourceType][id] = {
          type: resourceType,
          id,
          attributes,
          links,
          relationships
        };
        break;
      case "MERGE_RESOURCES":
        _initializeResource(draft, resourceType);

        Object.entries(resourcesById).forEach(
          ([id, resource]) => (draft[resourceType][id] = resource)
        );
        break;
      case "REMOVE_RESOURCE_BY_ID":
        delete draft[resourceType][id];
        break;
      case "CLEAR_RESOURCES":
        resourceTypes.forEach(resourceType => {
          draft[resourceType] = {};
        });
        break;
    }
  });
}

const _initializeResource = (draft, resourceType) => {
  if (resourceType in draft) return;
  draft[resourceType] = {};
};
