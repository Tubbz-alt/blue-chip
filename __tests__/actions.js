import {
  updateResources,
  updateResource,
  removeResource
} from "../lib/actions";
import jsonApiPayload
  from "../__testHelpers__/fixtrues/checklistsJsonApiResponse";
import graphQlPayload
  from "../__testHelpers__/fixtrues/checklistsGraphQlResponse";
import normalizedJsonApiTasksPayload
  from "../__testHelpers__/fixtrues/normalizedJsonApiTasksPayload";
import normalizedJsonApiChecklistsPayload
  from "../__testHelpers__/fixtrues/normalizedJsonApiChecklistsPayload";
import normalizedGraphQlTaskPayload
  from "../__testHelpers__/fixtrues/normalizedGraphQlTasksPayload";
import normalizedGraphQlChecklistPayload
  from "../__testHelpers__/fixtrues/normalizedGraphQlChecklistsPayload";
import resourcesReducer from "../src/resourcesReducer";

describe("actions", () => {
  describe("updateResources", () => {
    describe("JsonApi", () => {
      test("dispatches MERGE_RESOURCES for each ", () => {
        const dispatch = jest.fn();

        const tasksMergeResourcesAction = {
          resourceType: "tasks",
          resourcesById: normalizedJsonApiTasksPayload,
          type: "MERGE_RESOURCES"
        };

        const checklistsMergeResourcesAction = {
          resourceType: "checklists",
          resourcesById: normalizedJsonApiChecklistsPayload,
          type: "MERGE_RESOURCES"
        };

        updateResources(jsonApiPayload, dispatch);
        expect(dispatch).toBeCalledWith(tasksMergeResourcesAction);
        expect(dispatch).toBeCalledWith(checklistsMergeResourcesAction);
        expect(dispatch).toMatchSnapshot();
      });
    });

    describe("GraphQl", () => {
      test("dispatches MERGE_RESOURCES for each ", () => {
        const dispatch = jest.fn();

        const tasksMergeResourcesAction = {
          resourceType: "tasks",
          resourcesById: normalizedGraphQlTaskPayload,
          type: "MERGE_RESOURCES"
        };

        const checklistsMergeResourcesAction = {
          resourceType: "checklists",
          resourcesById: normalizedGraphQlChecklistPayload,
          type: "MERGE_RESOURCES"
        };

        updateResources(graphQlPayload, dispatch);
        expect(dispatch).toBeCalledWith(tasksMergeResourcesAction);
        expect(dispatch).toBeCalledWith(checklistsMergeResourcesAction);
        expect(dispatch).toMatchSnapshot();
      });
    });
  });

  describe("updateResource", () => {
    test("dispatches update action for a single resource", () => {
      const dispatch = jest.fn();

      const checklist = {
        id: 1,
        type: "checklists",
        attributes: { name: "Onboarding Rest" },
        links: { self: "http://example.com/checklists/1" },
        relationships: {
          tasks: { data: [{ id: 1, type: "tasks" }, { id: 2, type: "tasks" }] }
        }
      };

      const updateAction = {
        type: "ADD_OR_REPLACE_RESOURCE_BY_ID",
        resourceType: checklist.type,
        id: checklist.id,
        attributes: checklist.attributes,
        links: checklist.links,
        relationships: checklist.relationships
      };

      updateResource(checklist, dispatch);
      expect(dispatch).toBeCalledWith(updateAction);
      expect(dispatch).toMatchSnapshot();
    });
  });

  describe("removeResource", () => {
    test("dispatches remove resource action", () => {
      const dispatch = jest.fn();

      const checklist = {
        id: 1,
        type: "checklists"
      };

      const deleteAction = {
        type: "REMOVE_RESOURCE_BY_ID",
        resourceType: checklist.type,
        id: checklist.id
      };

      removeResource(checklist, dispatch);
      expect(dispatch).toBeCalledWith(deleteAction);
      expect(dispatch).toMatchSnapshot();
    });
  });
});
