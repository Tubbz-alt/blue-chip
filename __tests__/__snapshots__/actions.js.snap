// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions Mobx JsonApi dispatches MERGE_RESOURCES for each  1`] = `
Object {
  "checklists": Object {
    "1": Object {
      "attributes": Object {
        "name": "Onboarding Rest",
      },
      "id": 1,
      "links": Object {
        "self": "http://example.com/checklists/1",
      },
      "relationships": Object {
        "tasks": Object {
          "data": Array [
            Object {
              "id": 1,
              "type": "tasks",
            },
            Object {
              "id": 2,
              "type": "tasks",
            },
          ],
          "links": Object {
            "related": "http://example.com/checklists/1/tasks",
            "self": "http://example.com/checklists/1/relationships/tasks",
          },
        },
      },
      "type": "checklists",
    },
    "2": Object {
      "attributes": Object {
        "name": "Project Audit Rest",
      },
      "id": 2,
      "links": Object {
        "self": "http://example.com/checklists/2",
      },
      "relationships": Object {
        "tasks": Object {
          "data": Array [
            Object {
              "id": 3,
              "type": "tasks",
            },
            Object {
              "id": 4,
              "type": "tasks",
            },
          ],
          "links": Object {
            "related": "http://example.com/checklists/2/tasks",
            "self": "http://example.com/checklists/2/relationships/tasks",
          },
        },
      },
      "type": "checklists",
    },
    "3": Object {
      "attributes": Object {
        "name": "QA Checklist Rest",
      },
      "id": 3,
      "links": Object {
        "self": "http://example.com/checklists/3",
      },
      "relationships": Object {
        "tasks": Object {
          "data": Array [
            Object {
              "id": 5,
              "type": "tasks",
            },
            Object {
              "id": 6,
              "type": "tasks",
            },
          ],
          "links": Object {
            "related": "http://example.com/checklists/3/tasks",
            "self": "http://example.com/checklists/3/relationships/tasks",
          },
        },
      },
      "type": "checklists",
    },
  },
  "tasks": Object {
    "1": Object {
      "attributes": Object {
        "description": "Onboarding REST 1",
      },
      "id": 1,
      "links": Object {
        "self": "http://example.com/tasks/1",
      },
      "type": "tasks",
    },
    "2": Object {
      "attributes": Object {
        "description": "Onboarding REST 2",
      },
      "id": 2,
      "links": Object {
        "self": "http://example.com/tasks/2",
      },
      "type": "tasks",
    },
    "3": Object {
      "attributes": Object {
        "description": "Project Audit Rest 3",
      },
      "id": 3,
      "links": Object {
        "self": "http://example.com/tasks/3",
      },
      "type": "tasks",
    },
    "4": Object {
      "attributes": Object {
        "description": "Project Audit Rest 4",
      },
      "id": 4,
      "links": Object {
        "self": "http://example.com/tasks/4",
      },
      "type": "tasks",
    },
    "5": Object {
      "attributes": Object {
        "description": "QA Checklist Rest 5",
      },
      "id": 5,
      "links": Object {
        "self": "http://example.com/tasks/5",
      },
      "type": "tasks",
    },
    "6": Object {
      "attributes": Object {
        "description": "QA Checklist Rest 6",
      },
      "id": 6,
      "links": Object {
        "self": "http://example.com/tasks/6",
      },
      "type": "tasks",
    },
  },
}
`;

exports[`actions Mobx clearResources clears the store for the provided resources 1`] = `
Object {
  "checklists": Object {},
  "tasks": Object {},
}
`;

exports[`actions Mobx removeResource clears the store for the provided resources 1`] = `
Object {
  "checklists": Object {
    "2": Object {},
  },
}
`;
